File: Problems/1079. Letter Tile Possibilities/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int numTilePossibilities(string tiles) {
        unordered_map<char, int> freq;
        for (char c : tiles) freq[c]++;

        return backtrack(freq);
    }
private:
    int backtrack(unordered_map<char, int>& freq) {
        int count = 0;
        for (auto& [ch, num] : freq) {
            if (num > 0) {
                count++;
                num--;
                count += backtrack(freq);
                num++;
            }
        }
        return count;
    }
};

File: Problems/1079. Letter Tile Possibilities/Solution.rs
impl Solution {
    pub fn num_tile_possibilities(tiles: String) -> i32 {
        // Create a frequency array for uppercase English letters
        let mut counts = [0_u8; 26];

        // Helper function for backtracking
        fn backtrack(counts: &mut [u8; 26]) -> i32 {
            let mut sum: i32 = 0;
            for i in 0..26 {
                if counts[i] == 0 {
                    continue;
                }
                // Choose the tile
                counts[i] -= 1;
                // Count the current sequence
                sum += 1;
                // Continue building sequences with the remaining tiles
                sum += backtrack(counts);
                // Backtrack: restore the tile count
                counts[i] += 1;
            }
            sum
        }

        // Populate the frequency array with tile counts
        tiles.bytes().for_each(|b| counts[(b - b'A') as usize] += 1);

        // Start the backtracking process
        backtrack(&mut counts)
    }
}